source/extensions/filters/network/mysql_proxy/mysql_terminal_filter.cc

```diff
@@ -0,0 +1,522 @@
+    return;
+  }
+  parent.stepLocalSession(MYSQL_REQUEST_PKT_NUM, MySQLSession::State::Req);
+  parent.stepRemoteSession(MYSQL_RESPONSE_PKT_NUM, MySQLSession::State::ReqResp);
+}
+
+void MySQLTerminalFilter::DownstreamEventHandler::onCommandResponse(CommandResponse&) {
+  ENVOY_LOG(error, "mysql filter: downstream onCommandResponse impossible callback is called");
+}
+
+void MySQLTerminalFilter::UpstreamEventHandler::onNewMessage(MySQLSession::State state) {
+  parent.onNewMessage(state);
+  // close connection when received message on state NotHandled.
+  if (state == MySQLSession::State::NotHandled || state == MySQLSession::State::Error) {
+    ENVOY_LOG(info, "connection closed due to unexpected state occurs on communication");
+    parent.closeRemote();
+  }
+}
+
+void MySQLTerminalFilter::UpstreamEventHandler::onServerGreeting(ServerGreeting& greet) {
+  parent.onServerGreeting(greet);
+  ENVOY_LOG(debug, "server {} send challenge", greet.getVersion());
+  auto auth_method = AuthHelper::authMethod(greet.getServerCap(), greet.getAuthPluginName());
+  if (auth_method != AuthMethod::OldPassword && auth_method != AuthMethod::NativePassword) {
+    // If server's auth method is not supported, use mysql_native_password as response, wait for
+    // auth switch.
+    auth_method = AuthMethod::NativePassword;
+  }
+  seed = greet.getAuthPluginData();
+  auto login = MessageHelper::encodeClientLogin(
+      auth_method, parent.upstream_username_, parent.upstream_password_, parent.connect_db_, seed);
+  login.setSeq(greet.getSeq() + 1);
+
+  parent.sendRemote(login);
+  parent.stepRemoteSession(greet.getSeq() + 2, MySQLSession::State::ChallengeResp41);
+}
+
+void MySQLTerminalFilter::UpstreamEventHandler::onClientLogin(ClientLogin&) {
+  ENVOY_LOG(error, "mysql filter: upstream onClientLogin impossible callback is called");
+}
+
+void MySQLTerminalFilter::UpstreamEventHandler::onAuthSucc() {
+  parent.downstream_event_handler_->onBothAuthSucc();
+  ready = true;
+  parent.stepRemoteSession(MYSQL_RESPONSE_PKT_NUM, MySQLSession::State::ReqResp);
+  if (parent.downstream_event_handler_->buffer.length()) {
+    parent.downstream_event_handler_->decoder->onData(parent.downstream_event_handler_->buffer);
+  }
+}
+
+void MySQLTerminalFilter::UpstreamEventHandler::onClientLoginResponse(
+    ClientLoginResponse& login_resp) {
+  parent.onClientLoginResponse(login_resp);
+
+  absl::optional<ErrMessage> err;
+
+  switch (login_resp.getRespCode()) {
+  case MYSQL_RESP_AUTH_SWITCH: {
+    auto& auth_switch = dynamic_cast<AuthSwitchMessage&>(login_resp);
+    if (!auth_switch.isOldAuthSwitch() &&
+        auth_switch.getAuthPluginName() != "mysql_native_password") {
+      ENVOY_LOG(info, "auth plugin {} is not supported", auth_switch.getAuthPluginName());
+      err = MessageHelper::defaultError(
+          "mysql proxy is failed to auth to upstream due to auth plugin is not support (now "
+          "support mysql_native_password, mysql_old_password)");
```

Comment:
mysql: authentication with upstream failed due to unsupported plugin. Only mysql_native_password and mysql_old_password are supported.

Can you add to the log what type of auth the upstream wanted?

Reply:
fixed.